src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1716,38 +1792,46 @@
   {:pre [(instance? org.joda.time.DateTime report-ttl)
          (instance? org.joda.time.DateTime resource-events-ttl)
          (ifn? update-lock-status)]}
-  ;; We update these tables in a different order than report insertion, creates
-  ;; the potential for deadlocks when two concurrent transactions have each
-  ;; acquired a portion of their locks. So first, we acquire locks on all the
-  ;; tables in the same order they are accessed during insertion. These match
-  ;; the locks that would implicitly be acquired by the operations we're going
-  ;; to perform: Access exclusive for reports and resource_events so we can
-  ;; drop partition tables, and row exclusive to update certnames.
-  (acquire-locks! {"certnames" "ROW EXCLUSIVE"
-                   "reports" "ACCESS EXCLUSIVE"
-                   "resource_events" "ACCESS EXCLUSIVE"})
-  ;; force a resource-events GC. prior to partitioning, this would have happened
-  ;; via a cascade when the report was deleted, but now we just drop whole tables
-  ;; of resource events.
-  (delete-resource-events-older-than! (if (before? report-ttl resource-events-ttl)
-                                        resource-events-ttl report-ttl)
-                                      incremental?
-                                      update-lock-status)
-  (prune-daily-partitions "reports" report-ttl incremental?
-                          update-lock-status :write-locking-reports)
-  ;; since we cannot cascade back to the certnames table anymore, go clean up
-  ;; the latest_report_id column after a GC
-  (jdbc/do-commands
-   ["UPDATE certnames SET latest_report_id = NULL"
-    "  WHERE certname IN"
-    "    (SELECT DISTINCT certnames.certname FROM certnames"
-    "       LEFT OUTER JOIN reports ON (reports.certname = certnames.certname)"
-    "       WHERE reports.id IS NULL)"]))
-
-
+  (let [effective-resource-events-ttl
+         (if (before? report-ttl resource-events-ttl) resource-events-ttl report-ttl)]
+    (if-not (detach-partitions-concurrently?)
+      ;; PG11
+      (delete-reports-older-than-in-pg-11! report-ttl effective-resource-events-ttl
+                                           incremental? update-lock-status)
+      ;; PG14+
+      (do
```

Comment:
So this is fixed to work on PG14, but may still be subtly broken as I alluded to in the commit. Since the detach happens outside of a transaction, I think a periodic-gc and a manual sweep-reports could collide. Also the dropping of the tables no longer as the update-lock-status call.

Reply:
The drop-partition-tables now uses update-lock-status and call-with-lock-timeout like partition-daily-partitions.

However I think we still have a problem with detach in overlapping gc. An 'alter table [if exists] foo detach foo_x' can't be run idempotently. The if exists is only for whether the foo exists, not whether foo_x is still a partition of foo. Consequently detach-daily-partitions is vulnerable to get-partition-names being out of sync with detaching the partitions.